File: v2-swiftpm/swiftpm/Auv3PluginPackage/Sources/FrameworkMain/Messaging/MessageTypes.swift
enum MessageFromUI {
  case log(timestamp: Double, logKind: String, message: String)
  case uiLoaded
  case beginEdit(_ paramKey: String)
  case endEdit(_ paramKey: String)
  case performEdit(_ paramKey: String, _ value: Float)
  case instantEdit(_ paramKey: String, _ value: Float)
  case noteOnRequest(_ noteNumber: Int)
  case noteOffRequest(_ noteNumber: Int)
  case loadFullParameters(parameters: [String: Float])
  //
  case rpcReadFileRequest(rpcId: Int, path: String, skipIfNotExists: Bool)
  case rpcWriteFileRequest(rpcId: Int, path: String, content: String, append: Bool)
  case rpcDeleteFileRequest(rpcId: Int, path: String)
  case rpcLoadStateKvsItemsRequest(rpcId: Int)
  case writeStateKvsItem(key: String, value: String)
  case deleteStateKvsItem(key: String)
}

enum MessageFromApp {
  case setParameter(paramKey: String, value: Float)
  case bulkSendParameters(parameters: [String: Float])
  case hostNoteOn(noteNumber: Int)
  case hostNoteOff(noteNumber: Int)
  case hostTempo(tempo: Float)
  case hostPlayState(isPlaying: Bool)
  //
  case rpcReadFileResponse(rpcId: Int, success: Bool, content: String)
  case rpcWriteFileResponse(rpcId: Int, success: Bool)
  case rpcDeleteFileResponse(rpcId: Int, success: Bool)
  case rpcLoadStateKvsItemsResponse(rpcId: Int, items: [String: String])
}

File: v2-swiftpm/frontend/src/bridge/message-types.ts
export type MessageFromUi =
  | { type: "log"; timeStamp: number; logKind: string; message: string }
  | { type: "uiLoaded" }
  | { type: "beginEdit"; paramKey: string }
  | { type: "performEdit"; paramKey: string; value: number }
  | { type: "endEdit"; paramKey: string }
  | { type: "instantEdit"; paramKey: string; value: number }
  | { type: "noteOnRequest"; noteNumber: number }
  | { type: "noteOffRequest"; noteNumber: number }
  | {
      type: "loadFullParameters";
      parameters: Record<string, number>;
    }
  //
  | {
      type: "rpcReadFileRequest";
      rpcId: number;
      path: string;
      skipIfNotExists: boolean;
    }
  | {
      type: "rpcWriteFileRequest";
      rpcId: number;
      path: string;
      content: string;
      append: boolean;
    }
  | { type: "rpcDeleteFileRequest"; rpcId: number; path: string }
  | { type: "rpcLoadStateKvsItems"; rpcId: number }
  | { type: "writeStateKvsItem"; key: string; value: string }
  | { type: "deleteStateKvsItem"; key: string };

export type MessageFromApp =
  | { type: "setParameter"; paramKey: string; value: number }
  | { type: "bulkSendParameters"; parameters: Record<string, number> }
  | { type: "hostNoteOn"; noteNumber: number; velocity: number }
  | { type: "hostNoteOff"; noteNumber: number }
  | { type: "standaloneAppFlag" }
  //
  | {
      type: "rpcReadFileResponse";
      rpcId: number;
      success: boolean;
      content: string;
    }
  | { type: "rpcWriteFileResponse"; rpcId: number; success: boolean }
  | { type: "rpcDeleteFileResponse"; rpcId: number; success: boolean }
  | {
      type: "rpcLoadStateKvsItemsResponse";
      rpcId: number;
      items?: Record<string, string>;
    };

export type CoreBridge = {
  sendMessage: (msg: MessageFromUi) => void;
  subscribe: (listener: (msg: MessageFromApp) => void) => () => void;
};
